sdk/python/tests/compiler/testdata/input_artifact_raw_value.py

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python3
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import sys
+from pathlib import Path
+
+sys.path.insert(0, __file__ + '/../../../../')
+
+import kfp.dsl as dsl
+
+def component_with_input_artifact(text):
+    '''A component that passes text as input artifact'''
+
+    text_input_path = '/inputs/text/data'
+    return dsl.ContainerOp(
+        name='component_with_input_artifact',
+        input_artifact_paths={'text': text_input_path},
```

Comment:
Why are input_artifact_paths and input_artifact_arguments needed? Can we field for all input artifacts?

Reply:
Why not expose full argo input_artifacts spec here?

Reply:
> Why not expose full argo input_artifacts spec here?

We are moving away from hard-coding any artifact storage information in the pipeline Argo YAML. This makes the pipelines more portable which is a common request from customers and will become more and more important. 
Argo is moving in the same direction where all artifact repository information is configured in cluster-side configMaps.

Generally I do not want to expose more than we currently need, especially features that can be hard to support with different artifact passing systems (volumes, Airflow, Docker, etc)

Let's look at at Argo's spec for input artifacts: https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/types.go#L308

* name - Supported
* path - Supported
* mode - Not desirable: Not requested at this moment. Hard to support on GCS, passive volumes etc.
* from - N/A: Not supported for containers
* ArtifactLocation - Not desirable to expose: We want to prevent hard-coding any artifact repository information at the component level. The same effect can be archived at the argument level instead.
* globalName - Not desirable: Not requested at this moment. We do not want components to directly affect any global state.
* ArchiveStrategy - Not desirable to expose: Too Argo-specific. Bad compatibility with other data storage systems.
* optional - Might be desirable.

So, apart from `name` and `path` which I'm already including in this PR, there is only one additional field that can be desirable - `optional`.

Would you like me to replace
```input_artifact_paths={'text': text_input_path},```
with
```input_artifacts=[InputArtifact(name='text', path=text_input_path)],```
?

I'll be happy to do so.

Reply:
Could you explain why input_artifact_arguments is needed? Why not combine it with input_artifact_paths into one? 

I still think that we should support full argo's artifact feature and ContainerOp is a low level API which is argo specific. I am fine to expose partial of the argo's feature for now, but I don't want to limit the interface to be too opinionated. Maybe we should call out a design meeting to discuss this.

Reply:
> Could you explain why input_artifact_arguments is needed? Why not combine it with input_artifact_paths into one?

They are two different concepts: "input declaration" and "argument passing". They have some common parts, but they also have differences
These two lines are not the same:
```
# This line has input declarations
def my_func(param1: int = 3, param2 : str = 'bla'):
```
```
# This line has argument passing
my_func(param1=10, param2='bar')
```

Argo has the same distinction:
Please check the relevant Argo examples: https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml

This is Argo input artifact definition:
```
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
```
This is Argo artifact argument passing:
```
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"
```
These are two separate concepts, two separate parts and Argo does not mix them.

How can you even mix them when the same template is used multiple times? In this case there is a single input definition section and multiple argument passing sections. How can you merge them together?

>I still think that we should support full argo's artifact feature

Each added feature requires work.
Supporting `name` and `path` is crucial and also easy to implement (this PR).
Supporting other features is mach harder. It also limits our future options.

How would you support the `from` attribute in `ContainerOp`?

>I still think that we should support full argo's artifact feature

The real-world effect of this phrase is that the Pipelines project has zero support for **any** Argo artifact features for more than a year.
Perfect is the enemy of good. Let's start small and add more features along the way if we decide that they're needed.

Reply:
Generally I prefer the feedback to be actionable.
Also It's nice to have small PRs, so it's good when acting upon feedback does not increase the scope and size of the PR. We had cases where reviewers' requests have changed a small and focused PR to a 5000+ lines monster. It would be really great if we can keep the PR scopes as small as possible.

Reply:
I am confused by the interface, so I asked the question about the diff between input_artifact_arguments and input_artifact_paths and why do you need both? I also suggested to have a design discussion about the confusion here. I am not sure why you think it's non-actionable feedback.

Now I understand the diff from you last comment, but I still think it's unnecessary in DSL. Why do you need user to provide the name of the input artifact and duplicate it every time when they declare an input? The same reason applies why we don't ask user to declare each input parameters when they define the ContainerOp. For example, you can combine them into one field like following:

```python
    return dsl.ContainerOp(
        name='component_with_input_artifact',
        input_artifacts={ '/path/to/artifact1': artifact_value},
```
The name of the input artifact can be auto-generated. When converting to argo template, the compiler can take care of generating the template and argument parts separately.

Reply:
>For example, you can combine them into one field like following:

Thank you for actionable suggestion.

>The name of the input artifact can be auto-generated.

I'd rather prefer that the path is auto-generated:
```
input_artifacts={ 'artifact1': artifact_value},
```
How does that look to you?

Should the `input_artifacts` be a list of a dictionary?
Should we make some data type to describe the input artifact entry?
E.g.
```
input_artifacts=[InputArtifactArgument(path='/path/to/artifact1', argument=artifact_value)],
```

>The same reason applies why we don't ask user to declare each input parameters when they define the ContainerOp.

`ContainerOp` is a somewhat strange mixture of Argo's `ContainerTemplate` and `DagTask`. It has parts of both (although some parts are missing). This has already brought us some problems especially when Ning was implementing type checking. (To check argument types, you need to know the input types and for that you need inputs. Ning had to introduce new decorator and do some tricks just to recover the information about inputs.)

Reply:
> The name of the input artifact can be auto-generated. When converting to argo template, the compiler can take care of generating the template and argument parts separately.

This makes sense and I've implemented it. I even simplified it further (path and name can also be specified, but they're optional).

```
    return dsl.ContainerOp(
        name='consumer',
        image='alpine',
        command=['cat', dsl.InputArtifactArgument("Hello world")],
    )
```